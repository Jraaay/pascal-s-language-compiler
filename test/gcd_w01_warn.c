#include<stdio.h>
#include<stdbool.h>

int x, y; 

bool gcd(int a, int b){
	if(b == 0){
		return a; 
	}
	else{
		return gcd(b, a % b); 
	}
}

int main(int argc,  char* argv[]){
	scanf("%d%d", &x, &y); 
	printf("gcd(x, y): %d\n", gcd(x, y)); 
}
