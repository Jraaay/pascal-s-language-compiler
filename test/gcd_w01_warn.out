{
    "ast": {
        "type": "programstruct",
        "program_head": {
            "type": "program_head",
            "ID": "example",
            "idlist": {
                "type": "idlist",
                "ids": [
                    "input",
                    "output"
                ]
            }
        },
        "program_body": {
            "type": "program_body",
            "const_declarations": null,
            "var_declarations": {
                "type": "var_declarations",
                "var_declaration": {
                    "id": 0,
                    "type": "var_declaration",
                    "values": [
                        {
                            "idlist": {
                                "type": "idlist",
                                "ids": [
                                    "x",
                                    "y"
                                ]
                            },
                            "type": {
                                "type": "type",
                                "_type": {
                                    "type": "basic_type",
                                    "_type": "INTEGER"
                                }
                            }
                        }
                    ]
                }
            },
            "subprogram_declarations": {
                "type": "subprogram_declarations",
                "subprograms": [
                    {
                        "id": 4,
                        "type": "subprogram",
                        "subprogram_head": {
                            "type": "subprogram_head",
                            "_type": "FUNCTION",
                            "ID": "gcd",
                            "formal_parameter": {
                                "type": "formal_parameter",
                                "parameter_list": {
                                    "type": "parameter_list",
                                    "parameters": [
                                        {
                                            "type": "parameter",
                                            "value": {
                                                "type": "value_parameter",
                                                "idlist": {
                                                    "type": "idlist",
                                                    "ids": [
                                                        "a",
                                                        "b"
                                                    ]
                                                },
                                                "basic_type": {
                                                    "type": "basic_type",
                                                    "_type": "INTEGER"
                                                }
                                            }
                                        }
                                    ]
                                }
                            },
                            "basic_type": {
                                "type": "basic_type",
                                "_type": "BOOLEAN"
                            }
                        },
                        "subprogram_body": {
                            "type": "subprogram_body",
                            "const_declarations": null,
                            "var_declarations": [],
                            "compound_statement": {
                                "type": "compound_statement",
                                "statement_list": {
                                    "type": "statement_list",
                                    "statements": [
                                        {
                                            "type": "statement",
                                            "_type": "IF",
                                            "expression": {
                                                "type": "expression",
                                                "simple_expression_1": {
                                                    "type": "simple_expression",
                                                    "__type": "INTEGER",
                                                    "term": {
                                                        "type": "term",
                                                        "__type": "INTEGER",
                                                        "factor": {
                                                            "type": "factor",
                                                            "_type": "variable",
                                                            "__type": "INTEGER",
                                                            "variable": {
                                                                "type": "variable",
                                                                "__type": "INTEGER",
                                                                "ID": "b",
                                                                "id_varpart": {
                                                                    "type": "id_varpart",
                                                                    "expression_list": null
                                                                }
                                                            }
                                                        }
                                                    }
                                                },
                                                "RELOP": "=",
                                                "simple_expression_2": {
                                                    "type": "simple_expression",
                                                    "__type": "INTEGER",
                                                    "term": {
                                                        "type": "term",
                                                        "__type": "INTEGER",
                                                        "factor": {
                                                            "type": "factor",
                                                            "_type": "NUM",
                                                            "__type": "INTEGER",
                                                            "NUM": 0
                                                        }
                                                    }
                                                }
                                            },
                                            "statement": {
                                                "type": "statement",
                                                "_type": "variable",
                                                "variable": {
                                                    "type": "variable",
                                                    "__type": "BOOLEAN",
                                                    "ID": "gcd",
                                                    "id_varpart": {
                                                        "type": "id_varpart",
                                                        "expression_list": null
                                                    }
                                                },
                                                "ASSIGNOP": ":=",
                                                "expression": {
                                                    "type": "expression",
                                                    "__type": "INTEGER",
                                                    "simple_expression": {
                                                        "type": "simple_expression",
                                                        "__type": "INTEGER",
                                                        "term": {
                                                            "type": "term",
                                                            "__type": "INTEGER",
                                                            "factor": {
                                                                "type": "factor",
                                                                "_type": "variable",
                                                                "__type": "INTEGER",
                                                                "variable": {
                                                                    "type": "variable",
                                                                    "__type": "INTEGER",
                                                                    "ID": "a",
                                                                    "id_varpart": {
                                                                        "type": "id_varpart",
                                                                        "expression_list": null
                                                                    }
                                                                }
                                                            }
                                                        }
                                                    }
                                                }
                                            },
                                            "else_part": {
                                                "type": "else_part",
                                                "statement": {
                                                    "type": "statement",
                                                    "_type": "variable",
                                                    "variable": {
                                                        "type": "variable",
                                                        "__type": "BOOLEAN",
                                                        "ID": "gcd",
                                                        "id_varpart": {
                                                            "type": "id_varpart",
                                                            "expression_list": null
                                                        }
                                                    },
                                                    "ASSIGNOP": ":=",
                                                    "expression": {
                                                        "type": "expression",
                                                        "__type": "BOOLEAN",
                                                        "simple_expression": {
                                                            "type": "simple_expression",
                                                            "__type": "BOOLEAN",
                                                            "term": {
                                                                "type": "term",
                                                                "__type": "BOOLEAN",
                                                                "factor": {
                                                                    "type": "factor",
                                                                    "_type": "procedure_id",
                                                                    "__type": "BOOLEAN",
                                                                    "ID": "gcd",
                                                                    "expression_list": {
                                                                        "type": "expression_list",
                                                                        "__type": [
                                                                            "INTEGER",
                                                                            "INTEGER"
                                                                        ],
                                                                        "expressions": [
                                                                            {
                                                                                "type": "expression",
                                                                                "__type": "INTEGER",
                                                                                "simple_expression": {
                                                                                    "type": "simple_expression",
                                                                                    "__type": "INTEGER",
                                                                                    "term": {
                                                                                        "type": "term",
                                                                                        "__type": "INTEGER",
                                                                                        "factor": {
                                                                                            "type": "factor",
                                                                                            "_type": "variable",
                                                                                            "__type": "INTEGER",
                                                                                            "variable": {
                                                                                                "type": "variable",
                                                                                                "__type": "INTEGER",
                                                                                                "ID": "b",
                                                                                                "id_varpart": {
                                                                                                    "type": "id_varpart",
                                                                                                    "expression_list": null
                                                                                                }
                                                                                            }
                                                                                        }
                                                                                    }
                                                                                }
                                                                            },
                                                                            {
                                                                                "type": "expression",
                                                                                "__type": "INTEGER",
                                                                                "simple_expression": {
                                                                                    "type": "simple_expression",
                                                                                    "__type": "INTEGER",
                                                                                    "term": {
                                                                                        "type": "term",
                                                                                        "term": {
                                                                                            "type": "term",
                                                                                            "__type": "INTEGER",
                                                                                            "factor": {
                                                                                                "type": "factor",
                                                                                                "_type": "variable",
                                                                                                "__type": "INTEGER",
                                                                                                "variable": {
                                                                                                    "type": "variable",
                                                                                                    "__type": "INTEGER",
                                                                                                    "ID": "a",
                                                                                                    "id_varpart": {
                                                                                                        "type": "id_varpart",
                                                                                                        "expression_list": null
                                                                                                    }
                                                                                                }
                                                                                            }
                                                                                        },
                                                                                        "factor": {
                                                                                            "type": "factor",
                                                                                            "_type": "variable",
                                                                                            "__type": "INTEGER",
                                                                                            "variable": {
                                                                                                "type": "variable",
                                                                                                "__type": "INTEGER",
                                                                                                "ID": "b",
                                                                                                "id_varpart": {
                                                                                                    "type": "id_varpart",
                                                                                                    "expression_list": null
                                                                                                }
                                                                                            }
                                                                                        },
                                                                                        "__type": "INTEGER"
                                                                                    }
                                                                                }
                                                                            }
                                                                        ]
                                                                    }
                                                                }
                                                            }
                                                        }
                                                    }
                                                }
                                            }
                                        }
                                    ]
                                }
                            }
                        }
                    }
                ]
            },
            "compound_statement": {
                "type": "compound_statement",
                "statement_list": {
                    "type": "statement_list",
                    "statements": [
                        {
                            "type": "statement",
                            "_type": "READ",
                            "variable_list": {
                                "type": "variable_list",
                                "variables": [
                                    {
                                        "type": "variable",
                                        "__type": "INTEGER",
                                        "ID": "x",
                                        "id_varpart": {
                                            "type": "id_varpart",
                                            "expression_list": null
                                        }
                                    },
                                    {
                                        "type": "variable",
                                        "__type": "INTEGER",
                                        "ID": "y",
                                        "id_varpart": {
                                            "type": "id_varpart",
                                            "expression_list": null
                                        }
                                    }
                                ]
                            }
                        },
                        {
                            "type": "statement",
                            "_type": "WRITE",
                            "expression_list": {
                                "type": "expression_list",
                                "__type": [
                                    "BOOLEAN"
                                ],
                                "expressions": [
                                    {
                                        "type": "expression",
                                        "__type": "BOOLEAN",
                                        "simple_expression": {
                                            "type": "simple_expression",
                                            "__type": "BOOLEAN",
                                            "term": {
                                                "type": "term",
                                                "__type": "BOOLEAN",
                                                "factor": {
                                                    "type": "factor",
                                                    "_type": "procedure_id",
                                                    "__type": "BOOLEAN",
                                                    "ID": "gcd",
                                                    "expression_list": {
                                                        "type": "expression_list",
                                                        "__type": [
                                                            "INTEGER",
                                                            "INTEGER"
                                                        ],
                                                        "expressions": [
                                                            {
                                                                "type": "expression",
                                                                "__type": "INTEGER",
                                                                "simple_expression": {
                                                                    "type": "simple_expression",
                                                                    "__type": "INTEGER",
                                                                    "term": {
                                                                        "type": "term",
                                                                        "__type": "INTEGER",
                                                                        "factor": {
                                                                            "type": "factor",
                                                                            "_type": "variable",
                                                                            "__type": "INTEGER",
                                                                            "variable": {
                                                                                "type": "variable",
                                                                                "__type": "INTEGER",
                                                                                "ID": "x",
                                                                                "id_varpart": {
                                                                                    "type": "id_varpart",
                                                                                    "expression_list": null
                                                                                }
                                                                            }
                                                                        }
                                                                    }
                                                                }
                                                            },
                                                            {
                                                                "type": "expression",
                                                                "__type": "INTEGER",
                                                                "simple_expression": {
                                                                    "type": "simple_expression",
                                                                    "__type": "INTEGER",
                                                                    "term": {
                                                                        "type": "term",
                                                                        "__type": "INTEGER",
                                                                        "factor": {
                                                                            "type": "factor",
                                                                            "_type": "variable",
                                                                            "__type": "INTEGER",
                                                                            "variable": {
                                                                                "type": "variable",
                                                                                "__type": "INTEGER",
                                                                                "ID": "y",
                                                                                "id_varpart": {
                                                                                    "type": "id_varpart",
                                                                                    "expression_list": null
                                                                                }
                                                                            }
                                                                        }
                                                                    }
                                                                }
                                                            }
                                                        ]
                                                    }
                                                }
                                            }
                                        }
                                    }
                                ]
                            }
                        }
                    ]
                }
            }
        }
    },
    "symbolTable": {
        "constants": [],
        "variables": [
            {
                "id": 0,
                "token": "x",
                "type": "INTEGER",
                "isArray": false,
                "dimension": 0,
                "size": [],
                "start": [],
                "recordTable": null
            },
            {
                "id": 1,
                "token": "y",
                "type": "INTEGER",
                "isArray": false,
                "dimension": 0,
                "size": [],
                "start": [],
                "recordTable": null
            }
        ],
        "subFunc": [
            {
                "id": 4,
                "token": "gcd",
                "type": "BOOLEAN",
                "table": {
                    "params": 2,
                    "references": [
                        [
                            false,
                            false
                        ]
                    ],
                    "constants": [],
                    "variables": [
                        {
                            "id": 2,
                            "token": "a",
                            "type": "INTEGER",
                            "isArray": false,
                            "dimension": 0,
                            "size": [],
                            "start": [],
                            "recordType": null
                        },
                        {
                            "id": 3,
                            "token": "b",
                            "type": "INTEGER",
                            "isArray": false,
                            "dimension": 0,
                            "size": [],
                            "start": [],
                            "recordType": null
                        }
                    ]
                }
            }
        ]
    },
    "error": [
        {
            "code": "C-11",
            "info": {
                "line": 7,
                "value": "gcd",
                "lexpos": 159
            }
        }
    ],
    "warning": [
        {
            "code": "W-01",
            "info": {
                "line": 6,
                "value": [
                    "gcd",
                    "BOOLEAN",
                    "INTEGER"
                ],
                "lexpos": 132
            }
        }
    ]
}